File: packages/n2b/src/scanners/source.ts
import { applyBunApiRules } from "../rules/bun-apis";
import { applyNodeImportRules } from "../rules/node-imports";
import type { Finding, RunOptions } from "../types";
import { scanShebang } from "./shebang";

/**
 * Scan applicable à tout fichier source JS/TS.
 * Retourne les findings et le contenu après application des fixes autorisés
 * par `opts.mode` ('check' laisse intact, 'fix' applique safe, 'aggressive' applique tout).
 */
export function scanSource(
  path: string,
  content: string,
  opts: RunOptions,
): { findings: Finding[]; content: string } {
  const all: Finding[] = [];
  const aggressive = opts.mode === "aggressive";

  const shebang = scanShebang(path, content);
  all.push(...shebang.findings);
  let working = shebang.content;

  const nodeImp = applyNodeImportRules(path, working, aggressive);
  all.push(...nodeImp.findings);
  working = nodeImp.content;

  const bunApi = applyBunApiRules(path, working, aggressive);
  all.push(...bunApi.findings);
  working = bunApi.content;

  // En mode 'check', on ne modifie rien.
  if (opts.mode === "check") return { findings: all, content };

  return { findings: all, content: working };
}

File: rust/src/scanners/source.rs
use crate::rules::{bun_apis::apply_bun_api_rules, node_imports::apply_node_import_rules};
use crate::scanners::shebang::scan_shebang;
use crate::types::{Finding, Mode, RunOptions};

pub fn scan_source(path: &str, content: &str, opts: &RunOptions) -> (Vec<Finding>, String) {
    let mut all: Vec<Finding> = Vec::new();
    let aggressive = opts.mode == Mode::Aggressive;

    let (f, working) = scan_shebang(path, content);
    all.extend(f);
    let (f, working) = apply_node_import_rules(path, &working, aggressive);
    all.extend(f);
    let (f, working) = apply_bun_api_rules(path, &working, aggressive);
    all.extend(f);

    if opts.mode == Mode::Check {
        return (all, content.to_string());
    }
    (all, working)
}
